pkg/controller/elasticsearch/driver/upgrade_predicates.go

```diff
@@ -168,6 +168,102 @@ var predicates = [...]Predicate{
+		// This predicate checks that the following conditions are met for a candidate:
+		// * A cluster upgrade is in progress and the candidate version is not up to date
+		// * All primaries are assigned, only replicas are actually not assigned
+		// * There are no initializing or relocating shards
+		// See https://github.com/elastic/cloud-on-k8s/issues/1643
+		name: "if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			health, err := context.esState.Health()
+			if err != nil {
+				return false, err
+			}
+			_, healthyNode := context.healthyPods[candidate.Name]
+			if health != esv1.ElasticsearchYellowHealth || !healthyNode {
+				// This predicate is only relevant on healthy node if cluster health is yellow
+				return true, nil
+			}
+			version := candidate.Labels[label.VersionLabelName]
+			if version == context.es.Spec.Version {
+				// Restart in yellow state is only allowed during version upgrade
+				return false, nil
+			}
+			// This candidate needs a version upgrade, check if the Shards are in a compatible state.
+			shards, err := context.shardLister.GetShards()
+			if err != nil {
+				return false, err
+			}
+			for _, shard := range shards {
+				switch shard.State {
+				case client.INITIALIZING, client.RELOCATING:
+					return false, nil
+				case client.UNASSIGNED:
+					if !shard.IsReplica() {
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+		},
+	},
+	{
+		name: "do_not_delete_last_started_shard",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			allShards, err := context.shardLister.GetShards()
+			if err != nil {
+				return false, err
+			}
+			// We maintain two data structures to record:
+			// * The total number of replicas for a shard
+			// * How many of them are STARTED
+			startedReplicas := make(map[string]int)
+			replicas := make(map[string]int)
+			for _, shard := range allShards {
+				if shard.NodeName == candidate.Name {
+					continue
+				}
+				shardKey := shard.Key()
+				shardReplica := replicas[shardKey]
+				replicas[shardKey] = shardReplica + 1
+				if shard.State == client.STARTED {
+					assignedReplica := startedReplicas[shardKey]
+					startedReplicas[shardKey] = assignedReplica + 1
+				}
+			}
+
+			// Do not delete a node with a Primary if at least one replica is not STARTED
+			shardsByNode := allShards.GetShardsByNode()
+			shardsOnCandidate := shardsByNode[candidate.Name]
+			for _, shard := range shardsOnCandidate {
+				if !shard.IsPrimary() {
+					continue
+				}
+				shardKey := shard.Key()
+				replicas := replicas[shardKey]
+				assignedReplica := startedReplicas[shardKey]
+				if replicas > 0 && assignedReplica == 0 {
```

Comment:
It may be worthwhile to add a comment indicating that we accept there will be unavailability in the event that an index is configured with zero replicas